### PR TITLE
Feat: Repayment With Post Dated Checks

### PR DESCRIPTION
**Pull request by BLasan**
## Description

See [Issue-493](https://issues.apache.org/jira/browse/FINERACT-493)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by BLasan** [comment]
@thesmallstar Please take a look

**Comment by BLasan** [comment]
@thesmallstar Please review. Will open a separate PR for the integration test of repayment with post dated checks. Currently it's disabled

**Comment by BLasan** [comment]
@thesmallstar Please review

**Comment by BLasan** [comment]
@thesmallstar @ptuomola @luckyman20 Would you mind reviewing this?  Thanks :)

**Review by luckyman20 — COMMENTED** [review]
Code quality looks good to me.

**Comment by BLasan** [comment]
@ptuomola Could you please review this as well?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/repaymentwithpostdatedchecks/api/RepaymentWithPostDatedChecksApiResource.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    @Path("{installmentId}")
+    @Consumes({ MediaType.APPLICATION_JSON })
+    @Produces({ MediaType.APPLICATION_JSON })
+    @Operation(summary = "Get Post Dated Check", description = "Get Post Dated Check")
+    @ApiResponses({
+            @ApiResponse(responseCode = "200", description = "OK", content = @Content(array = @ArraySchema(schema = @Schema(implementation = PostDatedChecksApiResourceSwagger.GetPostDatedChecks.class)))) })
+    public String getPostDatedCheck(@PathParam("installmentId") @Parameter(description = "installmentId") final Integer installmentId,
+            @PathParam("loanId") @Parameter(description = "loanId") final Long loanId) {
+        /**
+         * TODO: Check the permission to read data.
```
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/repaymentwithpostdatedchecks/api/RepaymentWithPostDatedChecksApiResource.java]: Is this TODO still valid?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 2]
```diff
@@ -622,6 +647,7 @@ private void saveAndFlushLoanWithDataIntegrityViolationChecks(final Loan loan) {
         try {
             List<LoanRepaymentScheduleInstallment> installments = loan.getRepaymentScheduleInstallments();
             for (LoanRepaymentScheduleInstallment installment : installments) {
+                // installment.setPostDatedChecksToNull();
```
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:651]: Should this be removed?

File: fineract-provider/src/main/resources/sql/migrations/core_db/V27__add-loan-type-column-to-loan-table.sql [hunk 3]
```diff
@@ -19,48 +19,3 @@
 
 ALTER TABLE `m_loan`
     ADD COLUMN `loan_type_enum` SMALLINT NOT NULL AFTER `loan_status_id`;
-
--- create collateral management table
-CREATE TABLE IF NOT EXISTS `m_collateral_management` (
```
Comment (ptuomola) [inline fineract-provider/src/main/resources/sql/migrations/core_db/V27__add-loan-type-column-to-loan-table.sql]: Hmm... why are we deleting the lines relating to collateral management tables? Is this in the right PR?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/RepaymentWithPostDatedChecksTest.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    @BeforeEach
+    public void setup() {
+        Utils.initializeRESTAssured();
+        this.requestSpec = new RequestSpecBuilder().setContentType(ContentType.JSON).build();
+        this.requestSpec.header("Authorization", "Basic " + Utils.loginIntoServerAndGetBase64EncodedAuthenticationKey());
+        this.responseSpec = new ResponseSpecBuilder().expectStatusCode(200).build();
+    }
+
+    @Test
+    public void testRepaymentWithPostDatedChecks() {
```
Comment (ptuomola) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/RepaymentWithPostDatedChecksTest.java:65]: This seems to test only that the loan can be disbursed. Could you add integration tests for the other functionalities as well? i.e. repayment etc

**Review by ptuomola — CHANGES_REQUESTED** [review]

**Comment by ptuomola** [comment]
Hi @BLasan - looks good, left some comments. Can you have a look, and also rebase & squash please?

File: fineract-provider/src/main/resources/sql/migrations/core_db/V27__add-loan-type-column-to-loan-table.sql (hunk 3 above)
Comment (BLasan) [reply fineract-provider/src/main/resources/sql/migrations/core_db/V27__add-loan-type-column-to-loan-table.sql]: Nope. This collateral module is related to another PR and added the queries to a new file

**Comment by BLasan** [comment]
> Hi @BLasan - looks good, left some comments. Can you have a look, and also rebase & squash please?

Done. Please review

**Comment by BLasan** [comment]
@ptuomola Would you mind reviewing this as well? Thanks

**Review by ptuomola — APPROVED** [review]
LGTM

**Pull request merged**